app/notification.go

```diff
@@ -165,14 +165,20 @@ func (a *App) SendNotifications(post *model.Post, team *model.Team, channel *mod
 
 		umc := make(chan *model.AppError, 1)
 		go func(userId string) {
+			defer close(umc)
 			nErr := a.Srv().Store.Channel().IncrementMentionCount(post.ChannelId, userId)
 			if nErr != nil {
 				umc <- model.NewAppError("SendNotifications", "app.channel.increment_mention_count.app_error", nil, nErr.Error(), http.StatusInternalServerError)
-			} else {
-				umc <- nil
+				return
 			}
-
-			close(umc)
+			if *a.Config().ExperimentalSettings.ThreadAutoFollow && post.RootId != "" {
```

Comment:
I don't think this code belongs here. Looks like you are hijacking the async task to increment the mention count, to do other things. This method is already complex and hard to understand. I would prefer to separate it in its own go routing, and actually encapsulate that in a method to not contribute more to the growth of this method.

Reply:
Do you mean a separate go routine in this loop? The code is creating a thread membership for all people mentioned in a thread. It seems appropriate to put it in this loop.

Reply:
@jespino - I also think that the placement for this code is not that bad. This method does similar things, I'm just adding one more 'task'

Reply:
Yes, inside the loop, or later on looping through the `mentionedUsersList`. I don't know what is the best approach there. Probably makes no sense to fail the "notification" if the thread doesn't get followed anyway, so... maybe a different goroutine that does the looping and the execution.

Reply:
I disagree with that, the "updateMentionChans" looks like for update mentions count, not for threads autofollow, at least I would have a different name.

At the same time we are binding the success/failure of the incrementnotificationcount to the success or failure of the CreateMembershipIfNeeded, I don't know how much do we want that.

If you feel strong about it, I'm not against it, but looks like there are some rough edges there.